README.md
# MMM-JustForTodayMeditation

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

Narcotics Anonomous Just For Today Meditation Display for Magic Mirror2

![Preview](https://github.com/LeftBrainCreated/MMM-JustForTodayMeditation/blob/main/JustForToday_Cropped%20(1).png?raw=true)

Thanks to the BibleGateway module as it was used as a starting point.

Simple installation


Using Console, from /MagicMirror

`
cd modules
`

`
git clone https://github.com/LeftBrainCreated/MMM-JustForTodayMeditation.git
`

`
cd MMM-JustForTodayMeditation
`

`
npm i
`


## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:

```js
var config = {
  modules: [
		{
			module: 'MMM-JustForTodayMeditation',
			position: 'bottom_bar',
			config: {
				interval: 3 // refresh every x hours
				, retryDelay: 5000 // time in milliseconds
				, size: 'xsmall'
			}
		},
  ]
};
```

## Configuration options

| Option    | Description                                                                                                     |
| --------- | --------------------------------------------------------------------------------------------------------------- |
| `interval` | _Optional_ Refresh time, in hours Default is 3 hrs |
| `retryDelay` | _Optional_ When to try again, on failure |
| `size` | _Optional_ Size of text, default is |


CHANGELOG.md
# MMM-JustForTodayMeditation Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - Unreleased

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-JustForTodayMeditation
 *
 * By = LeftBrain
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");
var request = require("request");

var jftResult = "No Result Found";

module.exports = NodeHelper.create({
  // Subclass start method.
  start: function () {
    console.log("Started node_helper.js for MMM-JustForTodayMeditation.");
  },

  socketNotificationReceived: function (notification, payload) {
    console.log(
      this.name +
        " node helper received a socket notification: " +
        notification +
        " - Payload: " +
        payload
    );
    this.processJftRequest();
  },

  processJftRequest: function (version) {
    var self = this;

    var jftUrl = "http://www.jftna.org/jft/";

    request({ url: jftUrl, method: "GET" }, function (error, response, body) {
      switch (response.statusCode) {
        case 200:
          console.log("[JFT] RequestSuccess --> Response: " /* + body*/);

          // If you want the full read-out, uncomment this line 
          // and comment the next two below
          //var result = body;
          
          var regEx = new RegExp(/\<b\>Just for Today: \<\/b\>[^\<]*/g);
          var result = body.match(regEx);

          console.log("[JFT] " + result);

          self.sendSocketNotification("JFT_RESULT", result);
          break;

        default:
          if (error) {
            console.error(
              "[JFT] An error occurred while contacting jft webserver: " + error
            );
          } else {
            console.warn(
              "[JFT] Unknown Failure: \nResponseCode: " +
                response.statusCode +
                "\nResponseMessage: " +
                response.message
            );
          }
      }
    });
  },

  // Override socketNotificationReceived method.

  /* socketNotificationReceived(notification, payload)
   * This method is called when a socket notification arrives.
   *
   * argument notification string - The identifier of the noitication.
   * argument payload mixed - The payload of the notification.
   */
  socketNotificationReceived: function (notification, payload) {
    // if (notification === "MMM-JustForTodayMeditation-NOTIFICATION_TEST") {
    // 	console.log("Working notification system. Notification:", notification, "payload: ", payload);
    // 	// Send notification
    // 	this.sendNotificationTest(this.anotherFunction()); //Is possible send objects :)
    // }

    console.log(
      this.name +
        " node helper received a socket notification: " +
        notification +
        " - Payload: " +
        payload
    );
    this.processJftRequest();
  }

  // // Example function send notification test
  // sendNotificationTest: function(payload) {
  // 	this.sendSocketNotification("MMM-JustForTodayMeditation-NOTIFICATION_TEST", payload);
  // },

  // // this you can create extra routes for your module
  // extraRoutes: function() {
  // 	var self = this;
  // 	this.expressApp.get("/MMM-JustForTodayMeditation/extra_route", function(req, res) {
  // 		// call another function
  // 		values = self.anotherFunction();
  // 		res.send(values);
  // 	});
  // },

  // // Test another function
  // anotherFunction: function() {
  // 	return {date: new Date()};
  // }
});


MMM-JustForTodayMeditation.js
/* global Module */

/* Magic Mirror
 * Module: MMM-JustForTodayMeditation
 *
 * By =
 * MIT Licensed.
 */

Module.register("MMM-JustForTodayMeditation", {
  defaults: {
    interval: 1000 * 60 * 60 * 3, // every 3 hours
    retryDelay: 5000, // 5 seconds
    size: 'small'
  },

  requiresVersion: "2.1.0", // Required version of MagicMirror

  start: function () {
    var self = this;
    var jftResult = "";

    self.config.interval = (!self.config.interval) ? defaults.interval : self.config.interval;
    self.config.retryDelay = (!self.config.retryDelay) ? defaults.retryDelay : self.config.retryDelay;
    self.config.size = (!self.config.size) ? defaults.size : self.config.size;

    Log.info("Starting module: " + this.name);
    var self = this;

    //Do this once first
    self.sendSocketNotification("GET_JFT", null);

    //Then every hour
    setInterval(function () {
      self.sendSocketNotification("GET_JFT", null);
    }, this.config.interval * 1000 * 60 * 60); // count is in milliseconds, * 1000 for seconds, * 60 for minutes, * 60 for hours

  },

  /*
   * getData
   * function example return data and show it in the module wrapper
   * get a URL request
   *
   */
  getData: function () {
    var self = this;

    var urlApi = "https://jsonplaceholder.typicode.com/posts/1";
    var retry = true;

    var dataRequest = new XMLHttpRequest();
    dataRequest.open("GET", urlApi, true);
    dataRequest.onreadystatechange = function () {
      console.log(this.readyState);
      if (this.readyState === 4) {
        console.log(this.status);
        if (this.status === 200) {
          self.processData(JSON.parse(this.response));
        } else if (this.status === 401) {
          self.updateDom(self.config.animationSpeed);
          Log.error(self.name, this.status);
          retry = false;
        } else {
          Log.error(self.name, "Could not load data.");
        }
        if (retry) {
          self.scheduleUpdate(self.loaded ? -1 : self.config.retryDelay);
        }
      }
    };
    dataRequest.send();
  },

  /* scheduleUpdate()
   * Schedule next update.
   *
   * argument delay number - Milliseconds before next update.
   *  If empty, this.config.updateInterval is used.
   */
  scheduleUpdate: function (delay) {
    var nextLoad = this.config.updateInterval;
    if (typeof delay !== "undefined" && delay >= 0) {
      nextLoad = delay;
    }
    nextLoad = nextLoad;
    var self = this;
    setTimeout(function () {
      self.getData();
    }, nextLoad);
  },

  getDom: function () {
    console.log("[JFT] Updating DOM");

    // create element wrapper
    var wrapper = document.createElement("div");

    // value from config (size)
    if (this.jftResult) {
      var wrapper = document.createElement("div");
      switch (this.config.size) {
          case 'xsmall':
              wrapper.className = "bright xsmall";
              break;
          case 'small':
              wrapper.className = "bright small";
              break;
          case 'medium':
              wrapper.className = "bright medium";
              break;
          case 'large':
              wrapper.className = "bright large";
              break;
          default:
              wrapper.className = "bright medium";
      }
      wrapper.innerHTML = this.jftResult;
    }

    return wrapper;
  },

  getScripts: function () {
    return [];
  },

  getStyles: function () {
    return ["MMM-JustForTodayMeditation.css"];
  },

  processData: function (data) {
    var self = this;
    this.dataRequest = data;
    if (this.loaded === false) {
      self.updateDom(self.config.animationSpeed);
    }
    this.loaded = true;

    // the data if load
    // send notification to helper
    this.sendSocketNotification(
      "MMM-JustForTodayMeditation-NOTIFICATION_TEST",
      data
    );
  },

  // socketNotificationReceived from helper
  socketNotificationReceived: function (notification, payload) {
    switch (notification) {
      case "JFT_RESULT":
        this.jftResult = payload;
        this.updateDom();

        break;

      case "MMM-JustForTodayMeditation-NOTIFICATION_TEST":
        // set dataNotification
        this.dataNotification = payload;
        this.updateDom();

        break;

      default:
    }
  }
});
